### PR TITLE
Integrate metric items into raylet

### PR DESCRIPTION
**Pull request by jovany-wang**
## What do these changes do?
Integrate these metric items into raylet:
- [x] `RedisLatency`(Also, I refined the redis_context callback.)
- [x] `AvailableResource`
- [x] `TotalResource`
- [x] `CurrentDriver`
- [x] `ObjectManagerStats`
- [x] `ActorStats`
- [x]  `ConnectionPoolStats`
- [x] `LineageCacheStats`
- [x]  `ReconstructionPolicyStats`
- [x] `SchedulingQueueStats`
- [x]  `TaskDependencyManagerStats`

## Related issue number
https://github.com/ray-project/ray/issues/3858
## Linter

- [x]  I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: src/ray/gcs/redis_context.cc [hunk 1]
```diff
@@ -18,13 +21,24 @@ namespace {
...
-    }
+  RAY_CHECK(callback_index >= 0) << "The callback index must be greater than 0, "
+                                 << "but it actually be " << callback_index;
+  auto callback_item = ray::gcs::RedisCallbackManager::instance().get(callback_index);
+  if (!callback_item.is_subscription) {
+    // Record the redis latency for non-subscription redis operator.
+    auto end_time = current_sys_time_us();
+    RAY_LOG(DEBUG) << "Record a redis latency, start time is " << callback_item.start_time
+                   << "us, end time is " << end_time << "us, and the elapsed time is "
+                   << end_time - callback_item.start_time << "us.";
```
Comment (raulchen) [inline src/ray/gcs/redis_context.cc]: Is this debug log useful?

File: src/ray/gcs/redis_context.cc [hunk 2]
```diff
@@ -18,13 +21,24 @@ namespace {
...
-        ray::gcs::RedisCallbackManager::instance().get(callback_index)(data);
-    // Delete the callback if necessary.
-    if (delete_callback) {
-      ray::gcs::RedisCallbackManager::instance().remove(callback_index);
-    }
+  RAY_CHECK(callback_index >= 0) << "The callback index must be greater than 0, "
+                                 << "but it actually be " << callback_index;
+  auto callback_item = ray::gcs::RedisCallbackManager::instance().get(callback_index);
+  if (!callback_item.is_subscription) {
+    // Record the redis latency for non-subscription redis operator.
```
Comment (raulchen) [inline src/ray/gcs/redis_context.cc]:
```suggestion
    // Record the redis latency for non-subscription redis operations.
```

File: src/ray/gcs/redis_context.cc [hunk 3]
```diff
@@ -18,13 +21,24 @@ namespace {
...
 void ProcessCallback(int64_t callback_index, const std::string &data) {
-  if (callback_index >= 0) {
-    bool delete_callback =
-        ray::gcs::RedisCallbackManager::instance().get(callback_index)(data);
-    // Delete the callback if necessary.
-    if (delete_callback) {
-      ray::gcs::RedisCallbackManager::instance().remove(callback_index);
-    }
+  RAY_CHECK(callback_index >= 0) << "The callback index must be greater than 0, "
+                                 << "but it actually be " << callback_index;
```
Comment (raulchen) [inline src/ray/gcs/redis_context.cc]:
```suggestion
                                 << "but it actually is " << callback_index;
```

File: src/ray/gcs/redis_context.cc [hunk 4]
```diff
@@ -104,18 +118,20 @@ void SubscribeRedisCallback(void *c, void *r, void *privdata) {
...
+  auto start_time = current_sys_time_us();
+  callback_items_.emplace(num_callbacks_,
+                          CallbackItem(function, is_subscription, start_time));
   return num_callbacks_++;
 }
 
-RedisCallback &RedisCallbackManager::get(int64_t callback_index) {
-  RAY_CHECK(callbacks_.find(callback_index) != callbacks_.end());
-  return callbacks_[callback_index];
+RedisCallbackManager::CallbackItem RedisCallbackManager::get(int64_t callback_index) {
```
Comment (raulchen) [inline src/ray/gcs/redis_context.cc]:
```suggestion
RedisCallbackManager::CallbackItem& RedisCallbackManager::get(int64_t callback_index) const {
```

File: src/ray/gcs/redis_context.h [hunk 5]
```diff
@@ -19,9 +19,9 @@ namespace ray {
 
 namespace gcs {
 /// Every callback should take in a vector of the results from the Redis
-/// operation and return a bool indicating whether the callback should be
-/// deleted once called.
-using RedisCallback = std::function<bool(const std::string &)>;
+/// operation and return a bool indicating whether the callback is a subscription
+/// callback.
```
Comment (raulchen) [inline src/ray/gcs/redis_context.h]: This callback does't have return value now.

File: src/ray/raylet/lineage_cache.cc [hunk 6]
```diff
@@ -461,6 +461,7 @@ bool LineageCache::ContainsTask(const TaskID &task_id) const {
 
 const Lineage &LineageCache::GetLineage() const { return lineage_; }
 
+// TODO(qwang): Move records to `RecordMetrics` method.
```
Comment (raulchen) [inline src/ray/raylet/lineage_cache.cc]: it's already done?

File: src/ray/raylet/scheduling_queue.cc [hunk 7]
```diff
@@ -395,6 +408,22 @@ std::string SchedulingQueue::DebugString() const {
...
+           TaskState::WAITING,
+           TaskState::READY,
+           TaskState::RUNNING,
+           TaskState::INFEASIBLE,
+           TaskState::WAITING_FOR_ACTOR_CREATION,
+       }) {
+    stats::SchedulingQueueStats().Record(
+        static_cast<double>(GetTaskQueue(task_state)->GetTasks().size()),
+        {{stats::ValueTypeKey,
+          std::string("num_") + GetTaskStateString(task_state) + "tasks"}});
```
Comment (raulchen) [inline src/ray/raylet/scheduling_queue.cc]:
```suggestion
          std::string("num_") + GetTaskStateString(task_state) + "_tasks"}});
```

Comment (raulchen) [reply src/ray/raylet/scheduling_queue.cc]: does `GetTaskStateString` return upper case?

File: src/ray/stats/metric_defs.h [hunk 8]
```diff
@@ -12,16 +12,53 @@
 /// You can follow these examples to define your metrics.
 
 static Gauge CurrentWorker("current_worker",
-                           "This metric is used for report states of workers. "
+                           "This metric is used for report states of workers."
```
Comment (raulchen) [inline src/ray/stats/metric_defs.h]:
```suggestion
                           "This metric is used for reporting states of workers."
```

File: src/ray/stats/metric_defs.h [hunk 9]
```diff
@@ -12,16 +12,53 @@
...
                            "Through this, we can see the worker's state on dashboard.",
-                           "1 pcs", {NodeAddressKey, LanguageKey, WorkerPidKey});
+                           "1 pcs", {LanguageKey, WorkerPidKey});
+
+static Gauge CurrentDriver("current_driver",
+                           "This metric is used for report states of workers.", "1 pcs",
```
Comment (raulchen) [inline src/ray/stats/metric_defs.h]: The description isn't right?

File: src/ray/stats/metric_defs.h [hunk 10]
```diff
@@ -12,16 +12,53 @@
...
+                           {LanguageKey, DriverPidKey});
 
 static Count TaskCountReceived("task_count_received",
-                               "The count that the raylet received.", "pcs",
-                               {NodeAddressKey});
+                               "The count that the raylet received.", "pcs", {});
```
Comment (raulchen) [inline src/ray/stats/metric_defs.h]: `Number of tasks received by raylet.`

File: src/ray/raylet/node_manager.cc [hunk 11]
```diff
@@ -2188,9 +2209,18 @@ std::string NodeManager::DebugString() const {
...
   result << "\n- num dead actors: " << dead_actors;
   result << "\n- max num handles: " << max_num_handles;
+  stats::ActorStats().Record(live_actors,
+      {{stats::ValueTypeKey, "live_actors"}});
+  stats::ActorStats().Record(reconstructing_actors,
+      {{stats::ValueTypeKey, "reconstructing_actors"}});
+  stats::ActorStats().Record(dead_actors,
+      {{stats::ValueTypeKey, "dead_actors"}});
+  stats::ActorStats().Record(max_num_handles,
+      {{stats::ValueTypeKey, "max_num_handles"}});
```
Comment (raulchen) [inline src/ray/raylet/node_manager.cc]: why record the metrics in `DebugString()` function? Better use a separate function to do this.

File: src/ray/raylet/scheduling_queue.cc (hunk 7 above)
Comment (jovany-wang) [reply src/ray/raylet/scheduling_queue.cc]: It returns lowercases.

File: src/ray/gcs/redis_context.cc (hunk 1 above)
Comment (jovany-wang) [reply src/ray/gcs/redis_context.cc]: This log msg is used to check the latency metric result when I debuged, and it is not used any more. Let me remove it.

**Comment by jovany-wang** [comment]
I have address all comments, any more review suggestion?

File: src/ray/raylet/node_manager.cc [hunk 12]
```diff
@@ -2161,7 +2187,7 @@ void NodeManager::ForwardTask(const Task &task, const ClientID &node_id,
       });
 }
 
-void NodeManager::DumpDebugState() {
+void NodeManager::DumpDebugState() const {
```
Comment (raulchen) [inline src/ray/raylet/node_manager.cc]: Do we still need this function? `RecordMetrics` should do also report all the debugging information. @ericl

File: src/ray/raylet/node_manager.cc [hunk 13]
```diff
@@ -2213,6 +2226,40 @@ std::string NodeManager::DebugString() const {
   return result.str();
 }
 
+void NodeManager::RecordMetrics() const {
```
Comment (raulchen) [inline src/ray/raylet/node_manager.cc]: Who will call this function?
Also, please make sure this function doesn't execute when metrics isn't enabled.

File: src/ray/stats/metric_defs.h [hunk 14]
```diff
@@ -12,16 +12,53 @@
...
 
 static Gauge CurrentWorker("current_worker",
-                           "This metric is used for report states of workers. "
+                           "This metric is used for reporting states of workers."
                            "Through this, we can see the worker's state on dashboard.",
-                           "1 pcs", {NodeAddressKey, LanguageKey, WorkerPidKey});
+                           "1 pcs", {LanguageKey, WorkerPidKey});
+
+static Gauge CurrentDriver("current_driver",
+                           "This metric is used for report states of drivers.", "1 pcs",
```
Comment (raulchen) [inline src/ray/stats/metric_defs.h]:
```suggestion
                           "This metric is used for reporting states of drivers.", "1 pcs",
```

File: src/ray/stats/metric_defs.h [hunk 15]
```diff
@@ -12,16 +12,53 @@
...
-                               {NodeAddressKey});
+                               "Number of tasks received by raylet.", "pcs", {});
 
 static Histogram RedisLatency("redis_latency", "The latency of a Redis operation.", "us",
                               {100, 200, 300, 400, 500, 600, 700, 800, 900, 1000},
-                              {NodeAddressKey, CustomKey});
+                              {CustomKey});
+
+static Gauge LocalAvailableResource("local_available_resource",
+                                    "The available resource of this node.", "pcs",
```
Comment (raulchen) [inline src/ray/stats/metric_defs.h]:
```suggestion
                                    "The available resources on this node.", "pcs",
```

File: src/ray/stats/metric_defs.h [hunk 16]
```diff
@@ -12,16 +12,53 @@
...
+                                    {ResourceNameKey});
+
+static Gauge LocalTotalResource("local_total_resource",
+                                "The total resource of this node.", "pcs",
```
Comment (raulchen) [inline src/ray/stats/metric_defs.h]:
```suggestion
                                "The total resources on this node.", "pcs",
```

File: src/ray/stats/metric_defs.h [hunk 17]
```diff
@@ -12,16 +12,53 @@
...
+                                {ResourceNameKey});
+
+static Gauge ActorStats("actor_stats", "Stat the metric values of actor in raylet.",
```
Comment (raulchen) [inline src/ray/stats/metric_defs.h]:
```suggestion
static Gauge ActorStats("actor_stats", "Stat metrics of the actors in raylet.",
```
Also apply to the following fields

File: src/ray/stats/metric_defs.h [hunk 18]
```diff
@@ -12,16 +12,53 @@
...
+                        "pcs", {ValueTypeKey});
+
+static Gauge ObjectStats("object_stats", "Stat the metric values of object in raylet",
```
Comment (raulchen) [inline src/ray/stats/metric_defs.h]:
```suggestion
static Gauge ObjectManagerStats("object_stats", "Stat the metric values of object in raylet",
```

File: src/ray/raylet/node_manager.cc [hunk 19]
```diff
@@ -34,6 +34,32 @@ int64_t GetExpectedTaskCounter(
   return expected_task_counter;
 };
 
+struct ResultItem {
```
Comment (raulchen) [inline src/ray/raylet/node_manager.cc]: `ResultItem` is too general, consider renaming to sth like `ActorStats`

File: src/ray/raylet/node_manager.cc (hunk 12 above)
Comment (robertnishihara) [reply src/ray/raylet/node_manager.cc]: Let's keep it for now.

File: src/ray/raylet/node_manager.cc (hunk 13 above)
Comment (jovany-wang) [reply src/ray/raylet/node_manager.cc]: I think we shouldn't consider whether disable metrics here. The reasons are:
1) We have already consider this when recording.
2) This is one of the way we record, and we also record metrics other ways.

Comment (raulchen) [reply src/ray/raylet/node_manager.cc]: But this function's cost might not be trivial, considering the loops and functions calls. So we do a `if(not_enabled) return;` here

**Review by raulchen — APPROVED** [review]

**Pull request merged**